### PR TITLE
[DeviceSanitizer] Support nullpointer detection & enable GPU tests

### PR DESCRIPTION
**Pull request by AllanZyne**
UR: https://github.com/oneapi-src/unified-runtime/pull/1914

### CODE REVIEW

File: sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// REQUIRES: linux
+// RUN-NOT: %{build} %device_asan_flags -O0 -g -o %t
+// RUN-NOT: %{run} not %t 2>&1 | FileCheck %s
```
Comment (yingcong-wu) [inline sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp:3]: Why are those two RUN disabled? If we don't need them, can we remove them with a TODO?

File: sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// RUN: %{run} not %t 2>&1 | FileCheck %s
+// RUN: %{build} %device_asan_flags -O2 -g -o %t
+// RUN: %{run} not %t 2>&1 | FileCheck %s
+
+#include <sycl/detail/core.hpp>
+
+int main() {
+  sycl::queue Q;
+  constexpr std::size_t N = 4;
+  int *array = 0;
```
Comment (yingcong-wu) [inline sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp:14]: please use `nullptr`.

File: sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp (hunk 1 above)
Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp]: enabled

File: sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp (hunk 2 above)
Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp]: done

**Comment by AllanZyne** [comment]
Maybe we need to wait for cpu&gpu driver upgrading to support "__devicelib_exit" builtin.

**Comment by AllanZyne** [comment]
cpu & gpu e2e are disabled?

**Comment by wenju-he** [comment]
> Maybe we need to wait for cpu&gpu driver upgrading to support "__devicelib_exit" builtin.

Recently upgraded gpu driver in https://github.com/intel/llvm/pull/14838 has support of "__devicelib_exit" builtin

**Comment by AllanZyne** [comment]
@intel/unified-runtime-reviewers, please review

File: sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  sycl::queue Q;
+  constexpr std::size_t N = 4;
+  int *array = nullptr;
+
+  Q.submit([&](sycl::handler &h) {
+    h.parallel_for<class MyKernel>(
+        sycl::nd_range<1>(N, 1), [=](sycl::nd_item<1> item) { array[0] = 0; });
+    Q.wait();
+  });
+  // CHECK: ERROR: DeviceSanitizer: null-pointer-access on Unknown Memory
```
Comment (yingcong-wu) [inline sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp:21]: Can we construct cases to check null ptr for different address space?

Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp:21]: If I remember correctly, "0" is valid for local memory, and maybe we need to handle private memory as well.

Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/nullpointer/nullpointer.cpp]: Test case for private memory was added, but can't enable now due to gfx driver issue

**Comment by AllanZyne** [comment]
Kindly ping @intel/unified-runtime-reviewers

**Review by omarahmed1111 — APPROVED** [review]

**Comment by omarahmed1111** [comment]
@intel/llvm-gatekeepers please merge, Thanks!

**Comment by steffenlarsen** [comment]
Approval still needed from @intel/dpcpp-sanitizers-review (e.g. @yingcong-wu )

**Review by yingcong-wu — APPROVED** [review]

**Comment by omarahmed1111** [comment]
@intel/llvm-gatekeepers This should be ready now. Please merge.

**Comment by steffenlarsen** [comment]
@AllanZyne - The following tests are failing in post-commit after this:
  SYCL :: AddressSanitizer/invalid-argument/out-of-bounds.cpp
  SYCL :: AddressSanitizer/invalid-argument/released-pointer.cpp
  SYCL :: AddressSanitizer/nullpointer/global_nullptr.cpp
Please address these ASAP or we will will have to revert the patch.

**Comment by yingcong-wu** [comment]
> @AllanZyne - The following tests are failing in post-commit after this: SYCL :: AddressSanitizer/invalid-argument/out-of-bounds.cpp SYCL :: AddressSanitizer/invalid-argument/released-pointer.cpp SYCL :: AddressSanitizer/nullpointer/global_nullptr.cpp Please address these ASAP or we will will have to revert the patch.

Hi @steffenlarsen , @AllanZyne is OOO for now, so I think you should revert the patch.

**Comment by steffenlarsen** [comment]
@yingcong-wu - Thank you for making me aware! Looking at the changes, I don't think we need a full revert. If I'm reading it correctly, it seems like we didn't enable DG2 testing before, so I think it would be better to just disable the failing tests. Note that global_nullptr.cpp just seems to be failing to compile, likely due to unlucky timing, so I aim to fix that too. Patch in https://github.com/intel/llvm/pull/15450.

**Comment by nsirgien** [comment]
>Thank you for making me aware! Looking at the changes, I don't think we need a full revert. If I'm reading it correctly, it seems like we didn't enable DG2 testing before, so I think it would be better to just disable the failing tests.

@steffenlarsen, as far as I remember our discussion in SYCL CI meetings - it was an conclusion and agreement what we won't disable failing tests without at least tracker being created in regards of this failures, as it described in [CMPLRLLVM-58142](https://jira.devtools.intel.com/browse/CMPLRLLVM-58142).

Was it done in this case? I maybe wrong, but I don't see any tracker being mention in this PR in regards of this disabled tests and I also do not see it in https://github.com/intel/llvm/pull/15450.

**Comment by steffenlarsen** [comment]
> > Thank you for making me aware! Looking at the changes, I don't think we need a full revert. If I'm reading it correctly, it seems like we didn't enable DG2 testing before, so I think it would be better to just disable the failing tests.
> 
> @steffenlarsen, as far as I remember our discussion in SYCL CI meetings - it was an conclusion and agreement what we won't disable failing tests without at least tracker being created in regards of this failures.
> 
> Was it done in this case? I maybe wrong, but I don't see any tracker being mention in this PR in regards of this disabled tests and I also do not see it in #15450.

Hi @nsirgien! You're right, I did not mention the trackers in the PR itself. However, they are mentioned in the test files above the line disabling them. The trackers are https://github.com/intel/llvm/issues/15449 and https://github.com/intel/llvm/issues/15453.

**Pull request merged**